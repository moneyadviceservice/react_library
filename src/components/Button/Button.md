### Buttons allow users to take actions, and make choices, with a single tap.
There are multiple button variants to choose from. You can use the `plain` variant to serve as base for your own custom buttons.

If `href` is passed as a prop, the button will render as an anchor.

```jsx
import { Button, Row, Col } from '@moneypensionservice/directories';

<>
<Row>
  <Col padding={{vertical: '20px'}} sizes={4}>
    <Button text='Default' alignSelf="center" />
  </Col>
  <Col padding={{vertical: '20px'}} sizes={4}>
    <Button disabled text='Default Disabled' alignSelf="center" />
  </Col>
  <Col padding={{vertical: '20px'}} sizes={4}>
    <Button primary text='Primary' alignSelf="center" />
  </Col>
</Row>
<Row>
  <Col padding={{vertical: '20px'}} sizes={4}>
    <Button primary disabled text='Primary Disabled' alignSelf="center" />
  </Col>
  <Col padding={{vertical: '20px'}} sizes={4}>
    <Button blog text='Blog' alignSelf="center" />
  </Col>
  <Col padding={{vertical: '20px'}} sizes={4}>
    <Button plain text='Plain' alignSelf="center" />
  </Col>
</Row>
</>
```
