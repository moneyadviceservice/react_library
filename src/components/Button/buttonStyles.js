import styled, { css } from 'styled-components'
import {
  resolveMedia,
  genericStyles,
  setStyle,
  borderStyle,
} from '../../utils/helpers'

const defaultStyle = css`
  /** border-bottom */
  ${({ border, theme }) =>
    !border &&
    borderStyle({
      side: 'bottom',
      color: theme.colors.button.borderBottomColor,
      size: 'medium',
    })}
  background: ${({ theme }) => theme.colors.button.background};
  color: ${({ theme }) => theme.colors.button.color};
  ${({ weight, theme }) =>
    !weight && `font-weight: ${theme.typography.fontMedium}`};

  &:focus,
  &:hover,
  &:active {
    ${({ background, theme }) =>
      !background &&
      `background-color: ${theme.colors.button.activeBackground};`}
    ${({ border, theme }) =>
      !border && `border-color: ${theme.colors.button.activeBorderColor};`}
  }

  &:active {
    ${({ border, theme }) =>
      !border &&
      `border-bottom-color: ${theme.colors.button.borderBottomColor};
        border-top: 4px solid ${theme.colors.button.activeBorderColor}`}
  }

  &[disabled] {
    ${({ background, theme }) =>
      !background &&
      `background-color: ${theme.colors.button.disabledBackground};`}
    ${({ border, theme }) =>
      !border && `border-color: ${theme.colors.button.disabledBorderColor};`}
  }
`

const primaryStyle = css`
  /** border-bottom */
  ${({ border, theme }) =>
    !border &&
    borderStyle({
      side: 'bottom',
      color: theme.colors.button.primary.borderBottomColor,
      size: 'medium',
    })}
  background: ${({ theme }) => theme.colors.button.primary.background};
  color: ${({ theme }) => theme.colors.button.primary.color};
  ${({ weight, theme }) =>
    !weight && `font-weight: ${theme.typography.fontMedium}`};

  &:focus,
  &:hover,
  &:active {
    ${({ background, theme }) =>
      !background &&
      `background-color: ${theme.colors.button.primary.activeBackground};`}
    ${({ border, theme }) =>
      !border &&
      `border-color: ${theme.colors.button.primary.activeBorderColor};`}
  }

  &:active {
    ${({ border, theme }) =>
      !border &&
      `border-bottom-color: ${theme.colors.button.primary.borderBottomColor};
        border-top: 4px solid ${theme.colors.button.primary.activeBorderColor}`}
  }

  &[disabled] {
    ${({ background, theme }) =>
      !background &&
      `background-color: ${theme.colors.button.primary.disabledBackground};`}
    ${({ border, theme }) =>
      !border &&
      `border-color: ${theme.colors.button.primary.disabledBorderColor};`}
  }
`

const blogStyle = css`
  ${({ background, theme }) =>
    !background && `background-color: ${theme.colors.button.blog.masBlue};`}
  color: ${({ theme }) => theme.colors.white};
  ${({ weight, theme }) =>
    !weight && `font-weight: ${theme.typography.fontHeavy};`}
  ${({ padding }) => !padding && `padding: 12px 13.5px;`}
  ${({ border }) => !border && `border-radius: 5px; border-bottom: none;`}
  text-transform: uppercase;
  font-size: 18px;
  line-height: 18px;

  &:hover {
    background-color: ${({ theme }) => theme.colors.button.blog.masBlueDark};
  }
`

const ButtonWrapper = styled.button`
  ${genericStyles}

  /** common styles */
  -webkit-font-smoothing: antialiased;
  text-shadow: 1px 1px transparent;
  text-decoration: none;
  outline: 0;
  transition: background 0.2s ease;

  /** defaults */
  ${({ blog, theme }) =>
    !blog &&
    `font-size: 16px;
      line-height: 24px;
      border-radius: ${theme.sizes.button.border.radius};`}
  /** padding */
  ${({ blog, theme }) =>
    !blog && setStyle('padding', theme.sizes.button.padding)}
  /** border */
  ${({ border, theme }) =>
    !border && borderStyle(theme.sizes.button.border.size)}

  ${resolveMedia.md`
    ${({ blog, padding, theme }) =>
      !blog && !padding && setStyle('padding', theme.sizes.button.paddingMd)};
  `}

  &:focus,
  &:hover,
  &:active {
    text-decoration: none;
  }

  &:hover :not([disabled]) {
    cursor: pointer;
  }

  &:active {
    ${({ blog, padding }) => !padding && !blog && 'padding-top: 12px;'};
    ${({ blog, border }) => !border && !blog && 'border-bottom: none;'};
  }

  &[disabled] {
    box-shadow: none;
    ${({ blog, theme }) =>
      !blog && `color: ${theme.colors.button.disabledColor};`};
  }

  /** prop styles */
  ${({ weight }) => weight && `font-weight: ${weight};`}

  /** conditional styles */
  ${({ blog, plain, primary }) => {
    if (!plain) {
      if (blog) {
        return blogStyle
      } else if (primary) {
        return primaryStyle
      } else {
        return defaultStyle
      }
    }
  }}
`

export { ButtonWrapper }
