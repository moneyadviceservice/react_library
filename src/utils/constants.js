export const breakpoints = {
  xs: 1,
  sm: 48, // 768px
  md: 64, // 1024px
  lg: 90, // 1440px
  xl: 120, // 1920px
}

export const dimensions = ['xs', 'sm', 'md', 'lg', 'xl']

export const mediaQuery = 'only screen'

export const ALIGN_SELF_MAP = {
  center: 'center',
  end: 'flex-end',
  start: 'flex-start',
  stretch: 'stretch',
}

export const SPACING_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
]

