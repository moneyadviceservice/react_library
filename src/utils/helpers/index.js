export * from './background'
export * from './responsive'
export * from './flexbox'
export * from './styles'
export * from './mediaQueries'
export * from './sizing'
export * from './refs'
