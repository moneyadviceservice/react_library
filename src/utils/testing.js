import { render, act } from '@testing-library/react'
import GlobalTheme from '../components/ThemeProvider'

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalTheme, ...options })

const waitRender = async (ui, options) => {
  const { container, ...rest } = customRender(ui, options)

  await act(async () => {
    await Promise.resolve(container)
  })

  return { container, ...rest }
}

export * from '@testing-library/react'

export { customRender as render, waitRender }
