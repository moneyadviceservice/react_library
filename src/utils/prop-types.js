import PropTypes from 'prop-types'
import { SPACING_SIZES } from './constants'

/** spacing prop.type */
const spacingProp = PropTypes.shape({
  bottom: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  horizontal: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  left: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  right: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  top: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  vertical: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
})

export const genericPropTypes = {
  a11yTitle: PropTypes.string,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  margin: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  padding: PropTypes.oneOfType([
    PropTypes.oneOf([...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  border: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
}

export function genericPropsDefaults({
  a11yTitle = null,
  alignSelf = null,
  margin = null,
  padding = null,
  border = null,
} = {}) {
  return {
    a11yTitle: a11yTitle,
    alignSelf: alignSelf,
    margin: margin,
    padding: padding,
    border: border,
  }
}
