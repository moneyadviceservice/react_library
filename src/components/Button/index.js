import React from 'react'
import PropTypes from 'prop-types'
import { genericPropTypes, genericPropsDefaults } from '../../utils/prop-types'
// wrapper
import { ButtonWrapper } from './buttonStyles'

function Button({
  as,
  a11yTitle,
  primary,
  text,
  onClick,
  disabled,
  type = 'button',
  children,
  margin,
  alignSelf,
  href,
  onBlur,
  onFocus,
  onMouseOut,
  onMouseOver,
  ...rest
}) {
  const domTag = !as && !disabled && href ? 'a' : as

  return (
    <ButtonWrapper
      as={domTag}
      aria-label={a11yTitle}
      primary={primary}
      onClick={onClick}
      disabled={disabled}
      type={!href ? type : undefined}
      margin={margin}
      alignSelf={alignSelf}
      href={href}
      {...rest}
    >
      {text}
      {children}
    </ButtonWrapper>
  )
}

// Documentation
Button.propTypes = {
  /** The DOM tag or react component to use for the element. */
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  /** Whether this is a blog button. This will apply a different set of styles. */
  blog: PropTypes.bool,
  /** If specified, the button will behave like an anchor tag. */
  href: PropTypes.string,
  /** Label text to place inside the button. */
  text: PropTypes.string,
  /** Whether this is a primary button. There should be at most one per page or screen. */
  primary: PropTypes.bool,
  /** Whether the button is disabled. */
  disabled: PropTypes.bool,
  /** On click event. This triggers specific function. */
  onClick: PropTypes.func,
  /** Button content. */
  children: PropTypes.node,
  /** The type of button. Sets the type to submit for the default button on forms. */
  type: PropTypes.oneOf(['submit', 'button', 'reset']),
  /** The font weight for the text inside the button */
  weight: PropTypes.number,
  ...genericPropTypes,
}

Button.defaultProps = {
  blog: false,
  disabled: false,
  primary: false,
  type: 'button',
  weight: null,
  ...genericPropsDefaults({ alignSelf: 'start' }),
}

/** @component */
export default Button
