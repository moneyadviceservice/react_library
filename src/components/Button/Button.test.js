import React from 'react'
import { render, fireEvent } from '../../utils/testing'
import Button from '../Button'

describe('<Button />', () => {
  describe('Snapshots', () => {
    it('should match default snapshot', () => {
      const { container } = render(<Button />)
      expect(container).toMatchSnapshot()
    })
    it('should match default disabled snapshot', () => {
      const { container } = render(<Button disabled />)
      expect(container).toMatchSnapshot()
    })
    it('should match primary snapshot', () => {
      const { container } = render(<Button primary />)
      expect(container).toMatchSnapshot()
    })
    it('should match primary disabled snapshot', () => {
      const { container } = render(<Button primary disabled />)
      expect(container).toMatchSnapshot()
    })
    it('should match blog snapshot', () => {
      const { container } = render(<Button blog />)
      expect(container).toMatchSnapshot()
    })
    it('should match plain snapshot', () => {
      const { container } = render(<Button plain />)
      expect(container).toMatchSnapshot()
    })
  })

  describe('text prop', () => {
    it('should display value as Button children', () => {
      const { container } = render(<Button text="Cancel" />)

      expect(container.querySelector('button').textContent).toBe('Cancel')
    })
  })

  describe('onClick prop', () => {
    it('should call onClick function when Button is clicked', () => {
      const onClickMock = jest.fn()
      const { getByText } = render(<Button onClick={onClickMock} text="ok" />)

      fireEvent.click(getByText('ok'))

      expect(onClickMock).toHaveBeenCalledTimes(1)
    })
  })
})
