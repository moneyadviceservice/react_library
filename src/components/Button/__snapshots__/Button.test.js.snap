// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Snapshots should match blog snapshot 1`] = `
<div>
  <button
    class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 hrYGQy"
    type="button"
  />
</div>
`;

exports[`<Button /> Snapshots should match default disabled snapshot 1`] = `
<div>
  <button
    class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 cEHaJL"
    disabled=""
    type="button"
  />
</div>
`;

exports[`<Button /> Snapshots should match default snapshot 1`] = `
<div>
  <button
    class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 cEHaJL"
    type="button"
  />
</div>
`;

exports[`<Button /> Snapshots should match plain snapshot 1`] = `
<div>
  <button
    class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 dBANDC"
    type="button"
  />
</div>
`;

exports[`<Button /> Snapshots should match primary disabled snapshot 1`] = `
<div>
  <button
    class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 gxEGFn"
    disabled=""
    type="button"
  />
</div>
`;

exports[`<Button /> Snapshots should match primary snapshot 1`] = `
<div>
  <button
    class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 gxEGFn"
    type="button"
  />
</div>
`;
